File: robby-backend/src/main/java/Action.java
import java.util.Random;

public enum Action {

    MoveNorth   (0) {
        @Override
        public ActionResult perform(Position position, Board board) {
            return this.move(position, board, Offset.NORTH);
        }
    },
    MoveSouth   (1) {
        @Override
        public ActionResult perform(Position position, Board board) {
            return this.move(position, board, Offset.SOUTH);
        }
    },
    MoveEast    (2) {
        @Override
        public ActionResult perform(Position position, Board board) {
            return this.move(position, board, Offset.EAST);
        }
    },
    MoveWest    (3) {
        @Override
        public ActionResult perform(Position position, Board board) {
            return this.move(position, board, Offset.WEST);
        }
    },
    StayPut     (4) {
        @Override
        public ActionResult perform(Position position, Board board) {
            return new ActionResult(position, 0);
        }
    },
    PickUpCan   (5) {
        @Override
        public ActionResult perform(Position position, Board board) {
            boolean hadCan = board.tryRemoveCan(position);
            int reward = hadCan ? 10 : -1;
            return new ActionResult(position, reward);
        }
    },
    MoveRandom  (6) {
        @Override
        public ActionResult perform(Position position, Board board) {
            Offset offset;
            do {
                offset = Offset.values()[this.random.nextInt(Offset.values().length)];
            } while (offset == Offset.CURRENT);
            return move(position, board, offset);
        }
    };

    protected ActionResult move(Position position, Board board, Offset offset) {
        Position newPosition = offset.from(position);
        boolean isWall = board.isWall(newPosition);
        int reward = isWall ? -5 : 0;
        newPosition = isWall ? position : newPosition;
        return new ActionResult(newPosition, reward);
    }

    private int code;
    protected final Random random;

    Action(int code) {

        this.code = code;
        this.random = new Random();
    }

    public int getCode() {
        return code;
    }

    public abstract ActionResult perform(Position position, Board board);
}

File: robby-ui/src/game/action.js
'use strict';
import {DIRECTIONS} from "./directions.js";

export const ACTIONS = {
    moveNorth: 0,
    moveSouth: 1,
    moveEast: 2,
    moveWest: 3,
    stayPut: 4,
    pickUpCan: 5,
    moveRandom: 6,

    perform: function (point, action, cans) {
        switch (action) {
            case this.moveNorth:
                return this.move(point, DIRECTIONS.North);
            case this.moveEast:
                return this.move(point, DIRECTIONS.East);
            case this.moveSouth:
                return this.move(point, DIRECTIONS.South);
            case this.moveWest:
                return this.move(point, DIRECTIONS.West);
            case this.stayPut:
                return {
                    'reward': 0,
                    'position': point
                };
            case this.pickUpCan:
                return {
                    'reward': cans.tryRemove(point) ? 10 : -1,
                    'position': point
                };
            case this.moveRandom:
                let direction = DIRECTIONS.getRandom();
                return this.move(point, direction);
            default:
                throw "Unexpected action " + action;
        }
    },

    move: function(point, direction) {
        let next = point.add(direction);
        let isWall = next.isWall();

        return isWall
            ? {
                'reward': -5,
                'position': point
            }
            : {
                'reward': 0,
                'position': next
            }
    },
};

ACTIONS.byCode = [];
ACTIONS.byCode[ACTIONS.moveNorth] = ACTIONS.moveNorth;
ACTIONS.byCode[ACTIONS.moveSouth] = ACTIONS.moveSouth;
ACTIONS.byCode[ACTIONS.moveEast] = ACTIONS.moveEast;
ACTIONS.byCode[ACTIONS.moveWest] = ACTIONS.moveWest;
ACTIONS.byCode[ACTIONS.stayPut] = ACTIONS.stayPut;
ACTIONS.byCode[ACTIONS.pickUpCan] = ACTIONS.pickUpCan;
ACTIONS.byCode[ACTIONS.moveRandom] = ACTIONS.moveRandom;